packages/pwa-buildpack/lib/WebpackTools/plugins/UpwardIncludePlugin.js

```diff
@@ -0,0 +1,145 @@
+const debug = require('../../util/debug').makeFileLogger(__filename);
+const path = require('path');
+const CopyPlugin = require('copy-webpack-plugin');
+const { walkObject } = require('walk-object');
+const { promisify } = require('util');
+const jsYaml = require('js-yaml');
+
+/**
+ * @description webpack plugin that merged UPWARD configurations and
+ * autodetects file assets relied on by those configurations
+ */
+class UpwardIncludePlugin {
+    constructor({ upwardDirs }) {
+        this.upwardDirs = upwardDirs;
+        this.definition = {};
+        debug('created with dirs: %s', upwardDirs);
+    }
+    apply(compiler) {
+        this.compiler = compiler;
+        const { inputFileSystem } = compiler;
+        this.fs = {
+            readFile: promisify(inputFileSystem.readFile.bind(inputFileSystem)),
+            stat: promisify(inputFileSystem.stat.bind(inputFileSystem))
+        };
+        const onRun = () => this.onRun();
+        compiler.hooks.beforeRun.tapPromise('UpwardIncludePlugin', onRun);
+        compiler.hooks.watchRun.tapPromise('UpwardIncludePlugin', onRun);
+    }
+    async onRun() {
+        const { context } = this.compiler.options;
+        this.assetMap = {
+            'upward.yml': {
```

Comment:
As a minor suggestion, consider using a `Map` for `this.assetMap` if we're going to

- [x] populate it with unknown keys
- [x] iterate over it
- [x] have keys that require quotes

Reply:
In Node I'm not as worried about this, because this object only briefly exists. It's okay if it goes megamorphic.